File: workshop/20190323-clean-architecture/sample-03.ts
import * as readline from 'readline';
import * as process from 'process';

class TaxEntity {
    private readonly taxRate: number = 0.08;

    public calcTaxPrice(price: number): number {
        const p = this.taxRate + 1;

        return Math.round(price * p);
    }
}

class TaxCalculateUseCaseInputData {
    public readonly price: number;

    constructor(price: number) {
        this.price = price;
    }
}

class TaxCalculateUseCaseOutputData {
    public readonly price: number;

    constructor(price: number) {
        this.price = price;
    }
}

interface ITaxCalculateUseCase {
    calcTaxPrice(inputData: TaxCalculateUseCaseInputData): TaxCalculateUseCaseOutputData;
}

class TaxCalculateUseCase implements ITaxCalculateUseCase {
    public calcTaxPrice(inputData: TaxCalculateUseCaseInputData): TaxCalculateUseCaseOutputData {
        const taxEntity = new TaxEntity();
        const price = inputData.price;
        const taxPrice = taxEntity.calcTaxPrice(price);

        return new TaxCalculateUseCaseOutputData(taxPrice);
    }
}

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

rl.on('line', (input: string) => {
    const useCase = new TaxCalculateUseCase();
    const price = Number(input);
    const inputData = new TaxCalculateUseCaseInputData(price);
    const outputData = useCase.calcTaxPrice(inputData);
    const taxPrice = outputData.price;
    console.log(`税込価格:${taxPrice}`);
});

File: workshop/20190323-clean-architecture/py-sample/sample-03.py
import sys
from abc import *

class TaxEntity:
    __taxRate = 0.08

    # @type_definition(int)
    def calcTaxPrice(self, price):
        print("TaxEntity")
        p = self.__taxRate + 1
        return round(price * p)

class TaxCalculateUseCaseInputData:
    __price: int = 0

    # @type_definition(int)
    def __init__(self, price):
        print("TaxCalculateUseCaseInputData")
        self.__price = price

    def getPrice(self):
        return self.__price

class TaxCalculateUseCaseOutputData:
    __price: int = 0

    # @type_definition(int)
    def __init__(self, price):
        print("TaxCalculateUseCaseOutputData")
        self.__price = price

    def getPrice(self):
        return self.__price

class ITaxCalculateUseCase(metaclass=ABCMeta):
    @abstractmethod
    def calcTaxPrice(self, inputData: TaxCalculateUseCaseInputData):
        print("ITaxCalculateUseCase")
        pass

class TaxCalculateUseCase(ITaxCalculateUseCase):
    def calcTaxPrice(self, inputData: TaxCalculateUseCaseInputData):
        print("TaxCalculateUseCase")
        taxEntity = TaxEntity()
        price = inputData.getPrice()
        taxPrice = taxEntity.calcTaxPrice(price)
        return TaxCalculateUseCaseOutputData(taxPrice)

for line in iter(sys.stdin.readline, ""):
    useCase = TaxCalculateUseCase()
    price = int(line)
    inputData = TaxCalculateUseCaseInputData(price)
    outputData = useCase.calcTaxPrice(inputData)
    taxPrice = outputData.getPrice()
    print('taxPrice:%d ' % taxPrice)

